consensus/src/dag/dag_state_sync.rs

```diff
@@ -0,0 +1,180 @@
+        let dag_reader = self.dag_store.read();
+        (dag_reader.highest_ordered_round().unwrap_or_default() < li.commit_info().round()
+            && !dag_reader.exists_by_round_digest(
+                li.commit_info().round(),
+                li.ledger_info().consensus_data_hash(),
+            ))
+            || dag_reader.highest_committed_round().unwrap_or_default() + 2 * DAG_WINDOW
+                < li.commit_info().round()
+    }
+
+    pub async fn try_sync_to_highest_ordered_anchor(
+        &self,
+        node: &CertifiedNodeWithLedgerInfo,
+    ) -> anyhow::Result<Option<Arc<RwLock<Dag>>>> {
+        // Check whether to actually sync
+        let commit_li = node.ledger_info();
+        if !self.need_sync_for_ledger_info(commit_li) {
+            return Ok(None);
+        }
+
+        let dag_fetcher = Arc::new(DagFetcher::new(
+            self.epoch_state.clone(),
+            self.network.clone(),
+            self.time_service.clone(),
+        ));
+
+        self.sync_to_highest_ordered_anchor(node, dag_fetcher).await
+    }
+
+    /// Note: Assumes that the sync checks have been done
+    pub async fn sync_to_highest_ordered_anchor(
+        &self,
+        node: &CertifiedNodeWithLedgerInfo,
+        dag_fetcher: Arc<impl TDagFetcher>,
+    ) -> anyhow::Result<Option<Arc<RwLock<Dag>>>> {
+        let commit_li = node.ledger_info();
+
+        // TODO: there is a case where DAG fetches missing nodes in window and a crash happens and when we restart,
+        // we end up with a gap between the DAG and we need to be smart enough to clean up the DAG before the gap.
+
+        // Create a new DAG store and Fetch blocks
+        let target_round = node.round();
+        let start_round = commit_li.commit_info().round().saturating_sub(DAG_WINDOW);
+        let sync_dag_store = Arc::new(RwLock::new(Dag::new(
+            self.epoch_state.clone(),
+            self.storage.clone(),
+            start_round,
+        )));
+        let bitmask = { sync_dag_store.read().bitmask(target_round) };
+        let request = RemoteFetchRequest::new(
+            self.epoch_state.epoch,
+            node.parents_metadata().cloned().collect_vec(),
+            bitmask,
+        );
+
+        let responders = node
+            .certificate()
+            .signatures()
+            .get_signers_addresses(&self.epoch_state.verifier.get_ordered_account_addresses());
+
+        match dag_fetcher
+            .fetch(request, responders, sync_dag_store.clone())
+            .await
+        {
+            Ok(_) => {},
+            Err(err) => {
+                error!("error fetching nodes {}", err);
+                return Err(err);
+            },
+        }
+
+        // State sync
+        self.state_computer.sync_to(commit_li.clone()).await?;
+
+        {
+            let mut dag_writer = sync_dag_store.write();
+            dag_writer.prune();
+            if let Some(node_status) = dag_writer.get_node_ref_mut_by_round_digest(
```

Comment:
this is not enough, we need recursively marking all previous committed anchors in the window and all causal history

Reply:
discussed with @sasha8 on this, I think the best way is probably to read the events from block metadata that contains previous anchor information

Reply:
Also, it is not enough to mark only the anchor, we should mark all the causal history.

Reply:
Also, we are not persisting it. Won't this be a problem in recovery? Should we persist before calling state sync?

Reply:
We don't persist in normal path as well. I think db will give us the latest ledger info from which we can always get this info back even after recovery.

Reply:
But how will we know what nodes were already committed after recovery?